Cgal corefinement

fdrmrc opened the pull request:
Just a draft to describe things a little. 
- The "Surface with edge"d part in the following image the is result of the corefinement test. It is just a Surface_mesh.
![intersection_hedra_cube](https://user-images.githubusercontent.com/50499769/167211007-90dc6373-7fcc-44fd-a4f0-744f0d3c7f1d.png)

-  Here, that precise surface is filled with tets (use medit visualize it). This has still to be added
![Screenshot 2022-05-06 at 22 23 04](https://user-images.githubusercontent.com/50499769/167211654-58f1b370-6493-4e91-852b-0cea4fabf642.png)

fdrmrc commented:
- [ ] Fixed output files
- [ ] Added boolean operations between Surface_meshes
- [ ] Added Quadrature formulas over volumetric region bounded by a Surface_mesh and simple test

So far this works in 3D only I'll try to make this work in order to compute Quadratures over the intersection of <2,2> objects.

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -25,12 +25,24 @@
 
 namespace CGALWrappers
 {
+  enum class BooleanOperation
+  {
+    only_corefinement = 1 << 0,
+    union_op          = 1 << 2,
+    intersection_op   = 1 << 3,
+  };
```
```suggestion
  enum class BooleanOperation
  {
    none             = 1 << 0,
    union            = 1 << 1,
    intersection = 1 << 2,
  };
```

luca-heltai replied:
Shalle we support also difference?

fdrmrc replied:
Sure, thanks.

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -25,12 +25,24 @@
 DEAL_II_NAMESPACE_OPEN
 
 namespace CGALWrappers
 {
```
Add a bit of documentation, with some examples from the CGAL web site, or computed with the tests.

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +78,50 @@ namespace CGALWrappers
+   * Performs corefinement and boolean operations on two deal.II cells and put
+   * the result in a CGAL::Surface_mesh object.
+   *
+   * Extensive examples can be found in the CGAL documentation at
+   * https://doc.cgal.org/latest/Polygon_mesh_processing/index.html#coref_coref_subsec
+   *
```
I would put here some examples compute with this function, similar to the one you have in the comments.

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +78,50 @@ namespace CGALWrappers
+   * Extensive examples can be found in the CGAL documentation at
+   * https://doc.cgal.org/latest/Polygon_mesh_processing/index.html#coref_coref_subsec
+   *
+   * @tparam CGALPointType
+   * @tparam dim
+   * @tparam spacedim
```
Remove the template paramaters comments.

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +78,50 @@ namespace CGALWrappers
+   * @tparam CGALPointType
+   * @tparam dim
+   * @tparam spacedim
+   * @param[in] cell1
+   * @param[in] cell2
+   * @param[out] outsm
```
```suggestion
   * @param[in] cell1 First input cell
   * @param[in] cell2 Second input cell
   * @param[out] outsm Output surface mesh
```

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +78,50 @@ namespace CGALWrappers
+   * @param[in] cell1
+   * @param[in] cell2
+   * @param[out] outsm
+   * @param[in] boolean_operation BooleanOperation::intersection_op and
+   * BooleanOperation::union_op perform intersection and union of cells,
+   * respectively. If not specified, this parameter is defaulted to
+   * BooleanOperation::simple_corefinement, and only a corefinement is performed
```
```suggestion
   * @param[in] boolean_operation BooleanOperation::intersection and
   * BooleanOperation::union perform intersection and union of cells,
   * respectively. If not specified, this parameter is defaulted to
   * BooleanOperation::none, and only a corefinement is performed
```

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +78,50 @@ namespace CGALWrappers
+
+  /**
+   * Performs corefinement and boolean operations on two deal.II cells and put
+   * the result in a CGAL::Surface_mesh object.
```
What happens if the intersection is empty?

fdrmrc replied:
You end up in the default case, with `Assert(res,ExcMessage("The boolean operation you provided doesn't make sense. Please check it."));`

luca-heltai replied:
Ok. Can you add a comment here? The other option would be to simply return an empty mesh.

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +78,50 @@ namespace CGALWrappers
+   */
+  template <typename CGALPointType, int dim, int spacedim>
+  void
+  corefine_and_compute_boolean_operation_dealii_cells(
```
```suggestion
  boolean_operation(
```

luca-heltai replied:
I would make this simpler...

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +78,50 @@ namespace CGALWrappers
+  /**
+   * Build a Quadrature<spacedim> formula to integrate over the volumetric
+   * region described by a CGAL::Surface_mesh, by filling the interior with
+   * simplices and collecting points and weights to initegrate over it.
```
What happens if the surface mesh is not closed?

fdrmrc replied:
I cannot think to a case where having a non-closed surface makes sense, after all we're intersecting deal.II cells in that function.

luca-heltai replied:
I know, but does the function check that surface mesh is closed or not? You should document what happens in that case, and if the function throws an exception, you should state this here.

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +78,50 @@ namespace CGALWrappers
+   * region described by a CGAL::Surface_mesh, by filling the interior with
+   * simplices and collecting points and weights to initegrate over it.
+   *
+   * @tparam spacedim
+   * @tparam CGALPointType
```
Remove parameters documentation

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +78,50 @@ namespace CGALWrappers
+   *
+   * @tparam spacedim
+   * @tparam CGALPointType
+   * @param sm
+   * @param degree
+   * @return Quadrature<spacedim>
```
Comments input and output

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +78,50 @@ namespace CGALWrappers
+   * @return Quadrature<spacedim>
+   */
+  template <int spacedim, typename CGALPointType, typename CGALTriangulation>
+  Quadrature<spacedim>
```
```suggestion
  Quadrature<CGALTriangulation::Space_dimension()>
```
I'm not sure if this works, but this would allow us to remove the template parameter.

fdrmrc replied:
That is not constexpr. I used `CGALPointType::Ambient_dimension::value`  is constexpr, this allows to remove `spacedim`, thanks!

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +78,50 @@ namespace CGALWrappers
+   */
+  template <int spacedim, typename CGALPointType, typename CGALTriangulation>
+  Quadrature<spacedim>
+  quadrature_inside_region(const CGAL::Surface_mesh<CGALPointType> &sm,
```
Do we like the name?

fdrmrc replied:
Honestly I don't, but I was not able to find a meaningful name. Maybe `collect_quadratures_inside_surface` ?

luca-heltai reviewed (changes requested):
Can you make the PR on top of cgal-triangulation, instead of on top of master?

luca-heltai commented:
Do you think we could make this more general by having it return a polygon soup? I would really love to have this as we did with parmoonolith, i.e., arbitrary cell VS arbitrary cell. Of course if it is not possible, then we simply need to specialize this for each possibility.

fdrmrc commented:
Thanks for the comments! I addressed them and added a bit of documentation, except for the non-closed surface case. 
For what concerns the Polygon soup, I'm looking into it right now.

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -25,12 +25,40 @@
+   */
+  enum class BooleanOperation
+  {
+    NONE         = 1 << 0,
+    UNION        = 1 << 2,
+    INTERSECTION = 1 << 3,
+    DIFFERENCE   = 1 << 4,
```
deal.II notation: lowercase. If `none` doesn't work, use `corefine` as you had at the beginning. Aso, each one should be commented (using the `///<` doxygen tag): 
~~~
union = 1<<2, ///< Compute the union of the input arguments
intersection = 1<<2, ///< Compute the intersection...
~~~

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +94,50 @@ namespace CGALWrappers
+   * Performs corefinement and boolean operations on two deal.II cells and put
+   * the result in a CGAL::Surface_mesh object. If the cells are disjoint and
+   * you try to perform a boolean operation which makes no sense, like their
+   * intersection, an exception is thrown.
```
Do we want to throw an exception, or simply return an empty mesh? In the end, an empty set is a valid result for an intersection from a mathematical standpoint.

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +94,50 @@ namespace CGALWrappers
+   * @param[in] degree The degree of the quadrature formula
+   * @param[out] tr A CGAL triangulation storing the convex_hull of the volume
+   * bounded by the Surface_mesh
+   * @return Quadrature<spacedim> Collection of Quadratures. The sum of the weights equals to the volume of the polygonal region bounded by sm
```
```suggestion
   * @return Quadrature<spacedim> Collection of Quadratures. The sum of the weights
   * equals to the volume of the polygonal region bounded by sm
```

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +94,50 @@ namespace CGALWrappers
+    CGAL::Surface_mesh<CGALPointType> &                         outsm,
+    const BooleanOperation &boolean_operation = BooleanOperation::NONE);
+
+  /**
```
You no longer have `spacedim` at your disposal. You should either document what `spacedim` is, or use directly `CGALPointType::Ambient_dimension::value` in place of spacedim in the documentation.

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -25,12 +25,40 @@
+   * cube and a tetrahedron, converted to a deal.II triangulation
+   *
+   * @image html intersection_cube_with_tetrahedron.png
+   *
```
Perfect!

Can you also add the same picture computin the union and the difference?

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +94,50 @@ namespace CGALWrappers
+   * the result in a CGAL::Surface_mesh object. If the cells are disjoint and
+   * you try to perform a boolean operation which makes no sense, like their
+   * intersection, an exception is thrown.
+   *
```
You should add a line in which you describe who you subtract from whom. I'm imagining you do cell1 \setminus cell2, but you should document this. 

Also in this case, you could have an empty result (i.e., cell1 is entirely included in cell2).

luca-heltai commented on include/deal.II/cgal/surface_mesh.h:
```diff
@@ -66,6 +94,50 @@ namespace CGALWrappers
+   * Build a Quadrature<spacedim> formula to integrate over the volumetric
+   * region described by a CGAL::Surface_mesh, by filling the interior with
+   * simplices and collecting points and weights to integrate over it.
+   *
```
Add a comment that explains what happens if the operation results in a empty set: you return an empty quadrature.

luca-heltai reviewed (changes requested):
Very nice!

luca-heltai commented:
There is an issue with the image:

~~~
[100%] Built target documentation
/home/runner/work/dealii/dealii/include/deal.II/cgal/surface_mesh.h:51: warning: image file intersection_cube_with_tetrahedron.png is not found in IMAGE_PATH: assuming external image.
Error: Process completed with exit code 1.
~~~

luca-heltai commented on source/cgal/surface_mesh.cc:
```diff
@@ -114,6 +116,105 @@ namespace CGALWrappers
+  boolean_operation(
+    const typename Triangulation<dim, spacedim>::cell_iterator &cell1,
+    const typename Triangulation<dim, spacedim>::cell_iterator &cell2,
+    CGAL::Surface_mesh<CGALPointType> &                         outsm,
```
Add mapping to the input argument list.

luca-heltai commented on source/cgal/surface_mesh.cc:
```diff
@@ -114,6 +116,105 @@ namespace CGALWrappers
+        case BooleanOperation::DIFFERENCE:
+          res = PMP::corefine_and_compute_difference(sm1, sm2, outsm);
+          break;
+        case BooleanOperation::NONE:
```
This is not going to produce anything visible on the outside. outsm is not going to be touched by this function, and sm1 and sm2 will be destroied at the end of the call. We should document what will happen in this case, i.e., we return sm1, i.e., the corefinement of the first cell, therefore here you need to do `outsm = std::move(sm1);`

luca-heltai commented on source/cgal/surface_mesh.cc:
```diff
@@ -114,6 +116,105 @@ namespace CGALWrappers
+
+  template <typename CGALPointType, typename CGALTriangulation>
+  Quadrature<CGALPointType::Ambient_dimension::value>
+  collect_quadratures_inside_surface(
```
maybe a name like `build_quadrature_on_convex_hull`? 

Also, remove GCALTriangulation from the input argument. Since you are only using it inside the function, just build it inside.

luca-heltai replied:
We should also provide a different version of this function for non-convex polyhedral meshes. Maybe adding a boolean argument like `integrate_on_convex_hull` which defaults to `false` that uses the algorithm you showed me yesterday, i.e., create a domain, then a mesh, then last lines.

luca-heltai commented on source/cgal/surface_mesh.cc:
```diff
@@ -114,6 +116,105 @@ namespace CGALWrappers
+    const unsigned int                       degree,
+    CGALTriangulation &                      tr)
+  {
+    constexpr unsigned int spacedim = 3;
```
```suggestion
    constexpr unsigned int spacedim = CGALPointType::Ambient_dimension::value;
```

luca-heltai commented on source/cgal/surface_mesh.cc:
```diff
@@ -114,6 +116,105 @@ namespace CGALWrappers
+    QGaussSimplex<spacedim>              quad(degree);
+    std::vector<dealii::Point<spacedim>> pts;
+    std::vector<double>                  wts;
+    for (const auto &f : tr.finite_cell_handles())
```
this is different if spacedim is 3 or 2
